includes/abstracts/abstract-wc-stripe-payment-gateway.php

```diff
@@ -1053,16 +1053,26 @@ public function check_source( $prepared_source ) {
 	 * @param stdClass $source Source information.
 	 */
 	public function save_source_to_order( $order, $source ) {
-		// Store source in the order.
-		$order = WC_Stripe_Order::to_instance( $order );
-
 		$customer_id = $source->customer ?? null;
 		$source_id   = $source->source ?? null;
-		if ( $customer_id ) {
-			$order->set_stripe_customer_id( $customer_id );
-		}
-		if ( $source_id ) {
-			$order->set_source_id( $source_id );
+
+		if ( $this->is_subscription( $order ) || $this->has_pre_order( $order->get_id() ) ) {
+			if ( $source->customer ) {
+				$order->update_meta_data( '_stripe_customer_id', $customer_id );
+			}
+
+			if ( $source->source ) {
+				$order->update_meta_data( '_stripe_source_id', $source_id );
+			}
```

Comment:
Nit: To me, it feels unexpected to look at one field in the `if` check and then use a separate value in the code. Can we use the same variable for the check and the set? I am worried that we could end up with subtle bugs in future, e.g. if we change the way we set up the local variables.
```suggestion
			if ( $customer_id ) {
				$order->update_meta_data( '_stripe_customer_id', $customer_id );
			}

			if ( $source_id ) {
				$order->update_meta_data( '_stripe_source_id', $source_id );
			}
```

Reply:
Good catch. I copied the block from the pre-regression version and mistakenly did not update the conditions to use the variables. I am going to update it 👀 

Update: done in [f839d20](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4222/commits/f839d206ee482044c180444ff69b9fbfce77b6d6)